src/Server/BaseServer.ts

```diff
@@ -246,46 +255,55 @@ export abstract class BaseServer implements IServer {
     this.programs.push(program);
   }
 
+  writeToScriptFile(path: ScriptFilePath, content: string) {
+    return this.#writeContentToFile(path, content, this.scripts, Script);
+  }
+
+  writeToTextFile(path: TextFilePath, content: string) {
+    return this.#writeContentToFile(path, content, this.textFiles, TextFile);
+  }
+
   /**
-   * Write to a script file
+   * Write to a Script or TextFile
    * Overwrites existing files. Creates new files if the script does not exist.
    */
-  writeToScriptFile(filename: ScriptFilePath, code: string): writeResult {
-    // Check if the script already exists, and overwrite it if it does
-    const script = this.scripts.get(filename);
-    if (script) {
-      // content setter handles module invalidation
-      script.content = code;
-      return { overwritten: true };
-    }
-
-    // Otherwise, create a new script
-    const newScript = new Script(filename, code, this.hostname);
-    this.scripts.set(filename, newScript);
-    return { overwritten: false };
+  writeToContentFile(path: ContentFilePath, content: string): writeResult {
+    const isTextFile = hasTextExtension(path);
+    if (isTextFile) return this.writeToTextFile(path, content);
+    return this.writeToScriptFile(path, content);
   }
 
-  // Write to a text file
-  // Overwrites existing files. Creates new files if the text file does not exist
-  writeToTextFile(textPath: TextFilePath, txt: string): writeResult {
-    // Check if the text file already exists, and overwrite if it does
-    const existingFile = this.textFiles.get(textPath);
-    // overWrite if already exists
+  #writeContentToFile<T extends ContentFilePath>(
+    path: T,
+    content: string,
+    files: Map<string, ContentFile>,
+    ContentFile: new (path: T, content: string, host?: string) => ContentFile,
```

Comment:
`ContentFile` is a type, this is reusing the same name as a constructor which gets awfully ambiguous (since the constructor is *not* creating a literal ContentFile, but rather a subtype of it.)

I'd call this parameter `ctor` instead.